group-coordinator/src/main/java/org/apache/kafka/coordinator/group/share/ShareCoordinatorService.java

```diff
@@ -18,18 +18,152 @@
 package org.apache.kafka.coordinator.group.share;
 
 import org.apache.kafka.common.config.TopicConfig;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
 import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.coordinator.group.Record;
+import org.apache.kafka.coordinator.group.metrics.CoordinatorRuntimeMetrics;
+import org.apache.kafka.coordinator.group.runtime.CoordinatorEventProcessor;
+import org.apache.kafka.coordinator.group.runtime.CoordinatorLoader;
+import org.apache.kafka.coordinator.group.runtime.CoordinatorRuntime;
+import org.apache.kafka.coordinator.group.runtime.CoordinatorShardBuilderSupplier;
+import org.apache.kafka.coordinator.group.runtime.MultiThreadedEventProcessor;
+import org.apache.kafka.coordinator.group.runtime.PartitionWriter;
 import org.apache.kafka.server.record.BrokerCompressionType;
+import org.apache.kafka.server.util.timer.Timer;
+import org.slf4j.Logger;
 
+import java.time.Duration;
 import java.util.Properties;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.IntSupplier;
 
 public class ShareCoordinatorService implements ShareCoordinator {
-  private final int numPartitions;
+  private volatile int numPartitions = -1;
```

Comment:
I can see what you've done here, but it's a bit odd. The GroupCoordinatorService has a comment which clarifies it a bit. I suggest you copy that too.